translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,382 @@
+
+		if ocSpan.Resource != nil {
+			// Add a separate ResourceSpans item just for this span since it
+			// has a different Resource.
+			separateRS := data.NewResourceSpans(
+				ocNodeResourceToInternal(nil, ocSpan.Resource),
+				[]*data.Span{destSpan},
+			)
+			resourceSpanList = append(resourceSpanList, separateRS)
+		} else {
+			// Otherwise add the span to the first ResourceSpans item.
+			spanPtrs = append(spanPtrs, destSpan)
+		}
+	}
+
+	resourceSpans.SetSpans(spanPtrs)
+
+	return data.NewTraceData(resourceSpanList)
+}
+
+func ocSpanToInternal(dest *data.Span, src *octrace.Span) {
+	events, droppedEventCount := ocEventsToInternal(src.TimeEvents)
+	links, droppedLinkCount := ocLinksToInternal(src.Links)
+
+	dest.SetTraceID(data.TraceIDFromBytes(src.TraceId))
+	dest.SetSpanID(data.SpanIDFromBytes(src.SpanId))
+	dest.SetTraceState(ocTraceStateToInternal(src.Tracestate))
+	dest.SetParentSpanID(data.SpanIDFromBytes(src.ParentSpanId))
+	dest.SetName(src.Name.GetValue())
+	dest.SetKind(ocSpanKindToInternal(src.Kind, src.Attributes))
+	dest.SetStartTime(internal.TimestampToUnixnano(src.StartTime))
+	dest.SetEndTime(internal.TimestampToUnixnano(src.EndTime))
+	dest.SetAttributes(ocAttrsToInternal(src.Attributes))
+	dest.SetEvents(events)
+	dest.SetDroppedEventsCount(droppedEventCount)
+	dest.SetLinks(links)
+	dest.SetDroppedLinksCount(droppedLinkCount)
+	dest.SetStatus(ocStatusToInternal(src.Status))
+}
+
+func ocStatusToInternal(ocStatus *octrace.Status) data.SpanStatus {
+	if ocStatus == nil {
+		return data.SpanStatus{}
+	}
+	return data.NewSpanStatus(data.StatusCode(ocStatus.Code), ocStatus.Message)
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToInternal(ocTracestate *octrace.Span_Tracestate) data.TraceState {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(entry.Key)
+		sb.WriteString("=")
+		sb.WriteString(entry.Value)
+	}
+	return data.TraceState(sb.String())
+}
+
+func ocAttrsToInternal(ocAttrs *octrace.Span_Attributes) data.Attributes {
+	if ocAttrs == nil {
+		return data.Attributes{}
+	}
+
+	attrCount := len(ocAttrs.AttributeMap)
+	attrMap := make(data.AttributesMap, attrCount)
+	if attrCount > 0 {
+		values := data.NewAttributeValueSlice(attrCount)
+		i := 0
+		for key, ocAttr := range ocAttrs.AttributeMap {
+			switch attribValue := ocAttr.Value.(type) {
+			case *octrace.AttributeValue_StringValue:
+				values[i].SetString(attribValue.StringValue.GetValue())
+
+			case *octrace.AttributeValue_IntValue:
+				values[i].SetInt(attribValue.IntValue)
+
+			case *octrace.AttributeValue_BoolValue:
+				values[i].SetBool(attribValue.BoolValue)
+
+			case *octrace.AttributeValue_DoubleValue:
+				values[i].SetDouble(attribValue.DoubleValue)
+
+			default:
+				str := "<Unknown OpenCensus attribute value type>"
+				values[i].SetString(str)
+			}
```

Comment:
Should you not add the `span.kind` attribute here? If you want to keep it here, please comment why.

Reply:
Span.Kind is handled in `ocSpanKindToInternal`. This handles the attributes only. What kind of comment are looking for?

Reply:
I mean if we had in the oc attributes `span.kind=producer` we will end up with the right kind in the internal as well as the attribute. Should we remove the attribute if we use that to set the correct kind?